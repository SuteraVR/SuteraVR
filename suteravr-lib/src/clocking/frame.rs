use std::{
    fmt::Debug,
    io::{self, BufWriter, Cursor, Write},
};

use bytes::Buf;

/// Clocking-Server で扱うフレームを表すトレイトです。
pub trait ClockingFrame: Sized + Send + Sync + Debug {
    type Context;
    const MINIMAL_FRAME_SIZE: usize;

    /// フレームの解析を行います。
    ///
    /// フレームが解析可能であれば、`Some(Self)`を返し、そうでない場合は`None`を返します。
    #[allow(async_fn_in_trait)]
    async fn parse_frame(cursor: &mut Cursor<&[u8]>, ctx: Self::Context) -> Option<Self> {
        if cursor.remaining() < Self::MINIMAL_FRAME_SIZE {
            return None;
        }
        Self::parse_frame_unchecked(cursor, ctx).await
    }

    /// フレームの解析を行います。
    ///
    /// **このメソッドは直接呼びだすべきではありません。**  
    /// 代わりに、`parse_frame`を呼び出してください。
    ///
    /// 以下は、`parse_frame_unchecked`を実装する際のための注釈です。
    ///
    /// ## 期待される動作
    /// 1. フレームの大きさが足りない場合は、直ちに `None` を返します。
    /// 2. フレームの大きさが足りている場合は、解析を行います
    /// 3. 解析に成功した場合は、`Some(Self)` を返します。
    ///    このとき、`cursor` は、フレームの終端まで読み捨てている必要があります。
    /// 4. 解析に失敗した場合は、できるかぎり早期に`None` を返します。
    ///
    /// この関数が呼び出された時点で、`cursor`は少なくとも `Self::MINIMAL_FRAME_SIZE` バイトのデータを保持していることが保証されるので、  
    /// それに配慮する必要ありません。
    #[allow(async_fn_in_trait)]
    async fn parse_frame_unchecked(cursor: &mut Cursor<&[u8]>, ctx: Self::Context) -> Option<Self>;

    /// フレームの書き込みを行います。
    #[allow(async_fn_in_trait)]
    async fn write_frame<W: ?Sized + Write>(
        &self,
        cursor: &mut BufWriter<W>,
        ctx: Self::Context,
    ) -> io::Result<()>;
}

pub struct SuteraHeader {
    pub message_length: u64,
}
