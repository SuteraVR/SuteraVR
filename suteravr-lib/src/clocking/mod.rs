use std::io::Cursor;

use bytes::{Buf, BytesMut};
use thiserror::Error;
use tokio::{io::AsyncReadExt, net::TcpStream};

use crate::clocking::traits::ClockingFrame;

use self::traits::MessageAuthor;

pub mod oneshot_headers;
pub mod sutera_header;
pub mod sutera_status;
pub mod traits;

#[derive(Error, Debug)]
pub enum ClockingFramingError {
    #[error(transparent)]
    IoError(#[from] std::io::Error),
    #[error("Connection reset by peer")]
    ConnectionReset,
}

enum ConnectionContext {
    None,
    WaitStatus(usize),
    WaitMessageType(usize),
    WaitContent(usize),
}

pub enum ClockingFrameUnit {
    SuteraHeader(sutera_header::SuteraHeader),
    SuteraStatus(sutera_status::SuteraStatus),
    OneshotHeaders(oneshot_headers::OneshotHeader),
    Content(Vec<u8>),
    Unfragmented(Vec<u8>),
}

pub struct ClockingConnection {
    stream: TcpStream,
    buffer: BytesMut,
    author: MessageAuthor,
    context: ConnectionContext,
}
impl ClockingConnection {
    /// 既存のストリームから新しいClockingConnectionを作成します。
    ///
    /// **authorには、名前の通り「メッセージの送信者」が格納されることに注意してください。**
    /// たとえば、サーバー側で動いている場合は、[`MessageAuthor::Client`]を`author`に指定する必要があります。`
    pub fn new(stream: TcpStream, author: MessageAuthor) -> Self {
        Self {
            stream,
            author,
            buffer: BytesMut::with_capacity(4096),
            context: ConnectionContext::None,
        }
    }

    pub async fn read_frame(&mut self) -> Result<Option<ClockingFrameUnit>, ClockingFramingError> {
        loop {
            if let Some(frame) = self.parse_frame().await? {
                return Ok(Some(frame));
            }

            if self.stream.read_buf(&mut self.buffer).await? == 0 {
                if self.buffer.is_empty() {
                    return Ok(None);
                } else {
                    return Err(ClockingFramingError::ConnectionReset);
                }
            }
        }
    }

    async fn parse_frame(&mut self) -> Result<Option<ClockingFrameUnit>, ClockingFramingError> {
        let mut buf = Cursor::new(&self.buffer[..]);
        match self.context {
            ConnectionContext::None => {
                let remaining = buf.remaining();

                // SuteraHeader (通信のはじまりの目印) を探す
                // この時点で、SuteraHeaderの最小サイズよりも小さい場合は、次のバッファも読む
                if remaining < sutera_header::SuteraHeader::MINIMAL_FRAME_SIZE {
                    return Ok(None);
                }

                // 先頭からSuteraHeaderが成立していれば文句なしでOK
                buf.set_position(0);
                if let Some(header) =
                    sutera_header::SuteraHeader::parse_frame_unchecked(&mut buf, &()).await
                {
                    self.buffer.advance(buf.position() as usize);
                    self.context =
                        ConnectionContext::WaitStatus(header.message_length as usize - remaining);
                    return Ok(Some(ClockingFrameUnit::SuteraHeader(header)));
                }

                // 処理がここまで流れている時点で、
                // 最後にフレームが成立してから直ちにヘッダーが来ていないから何かがおかしい
                //
                // ただ、一応次にどこかでSuteraHeaderが来るかもしれないので、
                // バッファのどこかにSuteraHeaderを検知できたらそれまでのところをUnfragmentedとする
                for i in 1..=(remaining - sutera_header::SuteraHeader::MINIMAL_FRAME_SIZE) {
                    buf.set_position(i as u64);
                    if (sutera_header::SuteraHeader::parse_frame_unchecked(&mut buf, &()).await)
                        .is_some()
                    {
                        return Ok(Some(ClockingFrameUnit::Unfragmented(
                            self.buffer.copy_to_bytes(i).to_vec(),
                        )));
                    }
                }

                // 認識されていない状態でバッファが増えつづけると危険なので、
                // 定期的にUnfragmentedとして処理する
                if remaining > 1024 {
                    Ok(Some(ClockingFrameUnit::Unfragmented(
                        self.buffer.copy_to_bytes(1024).to_vec(),
                    )))
                } else {
                    Ok(None)
                }
            }
            ConnectionContext::WaitStatus(_) => todo!(),
            ConnectionContext::WaitMessageType(_) => todo!(),
            ConnectionContext::WaitContent(_) => todo!(),
        }
    }
}
