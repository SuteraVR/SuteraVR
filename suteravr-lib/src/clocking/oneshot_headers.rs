use std::mem::size_of;

use bytes::Buf;
use enum_map::enum_map;
use enum_map::{Enum, EnumMap};
use once_cell::sync::Lazy;

use crate::messaging::id::MessageId;
use crate::util::search_from_enum;

use super::traits::{ClockingFrame, MessageAuthor};

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
#[allow(non_camel_case_types)]
pub enum OneshotTypes {
    Connection_HealthCheck_Push,
    Connection_HealthCheck_Pull,
    Authentication_Login_Pull,
    TextChat_SendMessage_Pull,
    VoiceChat_SubVoiceTopic_Pull,
    VoiceChat_UnsubVoiceTopic_Pull,
    VoiceChat_SubAllVoiceTopic_Pull,
}

#[derive(PartialEq, Debug, Clone, Copy)]
pub enum OneshotDirection {
    Push,
    Pull,
}

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
pub enum OneshotStep {
    Request,
    Response,
}

static ONESHOT_TYPES_MAP: Lazy<
    EnumMap<OneshotTypes, [u8; OneshotHeader::MESSAGE_TYPE_DISTINCTOR_SIZE]>,
> = Lazy::new(|| {
    enum_map! {
        OneshotTypes::Connection_HealthCheck_Push     => [0x00, 0x00, 0x00, 0x00],
        OneshotTypes::Connection_HealthCheck_Pull     => [0x00, 0x00, 0x00, 0x01],
        OneshotTypes::Authentication_Login_Pull       => [0x00, 0x01, 0x00, 0x00],
        OneshotTypes::TextChat_SendMessage_Pull       => [0x00, 0x03, 0x00, 0x00],
        OneshotTypes::VoiceChat_SubVoiceTopic_Pull    => [0x00, 0x03, 0x01, 0x00],
        OneshotTypes::VoiceChat_UnsubVoiceTopic_Pull  => [0x00, 0x03, 0x01, 0x01],
        OneshotTypes::VoiceChat_SubAllVoiceTopic_Pull => [0x00, 0x03, 0x01, 0x02],
    }
});

pub static ONESHOT_DIRECTION_MAP: Lazy<EnumMap<OneshotTypes, OneshotDirection>> = Lazy::new(|| {
    enum_map! {
        OneshotTypes::Connection_HealthCheck_Push     => OneshotDirection::Push,
        OneshotTypes::Connection_HealthCheck_Pull     => OneshotDirection::Pull,
        OneshotTypes::Authentication_Login_Pull       => OneshotDirection::Pull,
        OneshotTypes::TextChat_SendMessage_Pull       => OneshotDirection::Pull,
        OneshotTypes::VoiceChat_SubVoiceTopic_Pull    => OneshotDirection::Pull,
        OneshotTypes::VoiceChat_UnsubVoiceTopic_Pull  => OneshotDirection::Pull,
        OneshotTypes::VoiceChat_SubAllVoiceTopic_Pull => OneshotDirection::Pull,
    }
});

pub static ONESHOT_STEP_MAP: Lazy<
    EnumMap<OneshotStep, [u8; OneshotHeader::MESSAGE_STEP_DISTINCTOR_SIZE]>,
> = Lazy::new(|| {
    enum_map! {
        OneshotStep::Request  => [0x01, 0x00],
        OneshotStep::Response => [0x01, 0x01],
    }
});

#[derive(Debug, PartialEq, Clone)]
pub struct OneshotHeader {
    pub step: OneshotStep,
    pub message_type: OneshotTypes,
    pub message_id: MessageId,
}

impl OneshotHeader {
    pub const MESSAGE_STEP_DISTINCTOR_SIZE: usize = 2;
    pub const MESSAGE_TYPE_DISTINCTOR_SIZE: usize = 4;
}

impl ClockingFrame for OneshotHeader {
    type Context = MessageAuthor;

    const MIN_FRAME_SIZE: usize = Self::MESSAGE_STEP_DISTINCTOR_SIZE
        + size_of::<MessageId>()
        + Self::MESSAGE_TYPE_DISTINCTOR_SIZE;

    fn parse_frame_unchecked(
        cursor: &mut std::io::Cursor<&[u8]>,
        ctx: &Self::Context,
    ) -> Option<Self> {
        let step = [cursor.get_u8(), cursor.get_u8()];
        match search_from_enum(*ONESHOT_STEP_MAP, &step) {
            Some(step) => {
                let message_id = cursor.get_u64();
                let message_type = [
                    cursor.get_u8(),
                    cursor.get_u8(),
                    cursor.get_u8(),
                    cursor.get_u8(),
                ];
                let Some(message_type) = search_from_enum(*ONESHOT_TYPES_MAP, &message_type) else {
                    return None;
                };
                if ONESHOT_DIRECTION_MAP[message_type]
                    != match (ctx, step) {
                        (MessageAuthor::Client, OneshotStep::Request) => OneshotDirection::Pull,
                        (MessageAuthor::Client, OneshotStep::Response) => OneshotDirection::Push,
                        (MessageAuthor::Server, OneshotStep::Request) => OneshotDirection::Push,
                        (MessageAuthor::Server, OneshotStep::Response) => OneshotDirection::Pull,
                    }
                {
                    return None;
                }
                Some(Self {
                    step,
                    message_type,
                    message_id,
                })
            }
            None => None,
        }
    }

    async fn write_frame<W: tokio::io::AsyncWriteExt + Unpin>(
        &self,
        stream: &mut W,
        _ctx: &Self::Context,
    ) -> std::io::Result<()> {
        stream.write_all(&ONESHOT_STEP_MAP[self.step]).await?;
        stream.write_u64(self.message_id).await?;
        stream
            .write_all(&ONESHOT_TYPES_MAP[self.message_type])
            .await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::clocking::traits::test_util::{
        encode, test_clockingframe_mustfail, test_clockingframe_reflective,
    };

    use super::*;

    #[tokio::test]
    async fn clockingserver_oneshot_header() {
        test_clockingframe_reflective(
            OneshotHeader {
                step: OneshotStep::Request,
                message_type: OneshotTypes::Connection_HealthCheck_Push,
                message_id: 0x1234567890abcdef,
            },
            MessageAuthor::Server,
        )
        .await;
        test_clockingframe_reflective(
            OneshotHeader {
                step: OneshotStep::Request,
                message_type: OneshotTypes::Connection_HealthCheck_Pull,
                message_id: 0x1234567890abcdef,
            },
            MessageAuthor::Client,
        )
        .await;
        test_clockingframe_reflective(
            OneshotHeader {
                step: OneshotStep::Response,
                message_type: OneshotTypes::Connection_HealthCheck_Push,
                message_id: 0x1234567890abcdef,
            },
            MessageAuthor::Client,
        )
        .await;
        test_clockingframe_reflective(
            OneshotHeader {
                step: OneshotStep::Response,
                message_type: OneshotTypes::Connection_HealthCheck_Pull,
                message_id: 0x1234567890abcdef,
            },
            MessageAuthor::Server,
        )
        .await;
    }

    #[tokio::test]
    async fn clockingserver_oneshot_header_mismatch_direction_pull() {
        let payload = encode(
            &OneshotHeader {
                step: OneshotStep::Request,
                message_type: OneshotTypes::Connection_HealthCheck_Pull,
                message_id: 0x1234567890abcdef,
            },
            &MessageAuthor::Server,
        )
        .await;
        test_clockingframe_mustfail::<OneshotHeader>(
            &payload,
            &MessageAuthor::Server,
            Some(OneshotHeader::MIN_FRAME_SIZE),
        )
        .await;
    }

    #[tokio::test]
    async fn clockingserver_oneshot_header_mismatch_direction_push() {
        let payload = encode(
            &OneshotHeader {
                step: OneshotStep::Request,
                message_type: OneshotTypes::Connection_HealthCheck_Push,
                message_id: 0x1234567890abcdef,
            },
            &MessageAuthor::Client,
        )
        .await;
        test_clockingframe_mustfail::<OneshotHeader>(
            &payload,
            &MessageAuthor::Client,
            Some(OneshotHeader::MIN_FRAME_SIZE),
        )
        .await;
    }
}
