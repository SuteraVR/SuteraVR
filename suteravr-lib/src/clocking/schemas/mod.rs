pub mod event;
pub mod oneshot;
