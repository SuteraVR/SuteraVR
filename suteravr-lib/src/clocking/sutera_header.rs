use bytes::Buf;
use tokio::io::AsyncWriteExt;

use crate::messaging::version::Version;

use super::traits::ClockingFrame;

#[derive(Debug, PartialEq, Clone)]
pub struct SuteraHeader {
    pub version: Version,
}

impl SuteraHeader {
    const PREFIX: &'static [u8] = b"SuteraVR";
}
impl ClockingFrame for SuteraHeader {
    type Context = ();
    const MIN_FRAME_SIZE: usize = Self::PREFIX.len() + Version::MIN_FRAME_SIZE;

    async fn parse_frame_unchecked(
        cursor: &mut std::io::Cursor<&[u8]>,
        _ctx: &Self::Context,
    ) -> Option<Self> {
        if Self::PREFIX != cursor.copy_to_bytes(Self::PREFIX.len()) {
            return None;
        }
        let version = Version::parse_frame_unchecked(cursor, &()).await?;
        Some(Self { version })
    }

    async fn write_frame<W: tokio::io::AsyncWriteExt + Unpin>(
        &self,
        stream: &mut tokio::io::BufWriter<W>,
        _ctx: &Self::Context,
    ) -> std::io::Result<()> {
        stream.write_all(Self::PREFIX).await?;
        self.version.write_frame(stream, _ctx).await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::clocking::traits::test_util::{
        encode, test_clockingframe_mustfail, test_clockingframe_reflective,
    };

    use super::*;

    #[tokio::test]
    async fn clockingframe_suteraheader() {
        test_clockingframe_reflective(
            SuteraHeader {
                version: Version {
                    major: 1,
                    minor: 2,
                    patch: 3,
                },
            },
            (),
        )
        .await
    }

    #[tokio::test]
    async fn clockingframe_suteraheader_prefixerr() {
        let mut encoded = encode(
            &SuteraHeader {
                version: Version {
                    major: 1,
                    minor: 2,
                    patch: 3,
                },
            },
            &(),
        )
        .await;
        encoded[0] = b'T';
        encoded[1] = b'e';
        encoded[2] = b's';
        encoded[3] = b'u';
        encoded[4] = b'r';
        encoded[5] = b'a';

        test_clockingframe_mustfail::<SuteraHeader>(
            &encoded,
            &(),
            Some(SuteraHeader::PREFIX.len()),
        )
        .await;
    }
}
