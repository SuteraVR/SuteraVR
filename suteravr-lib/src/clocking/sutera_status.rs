use bytes::Buf;
use enum_map::enum_map;
use enum_map::Enum;
use enum_map::EnumMap;
use once_cell::sync::Lazy;

use crate::util::search_from_enum;

use super::traits::ClockingFrame;

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
pub enum SuteraStatusError {
    SchemaVersionNotSupported,
    BadRequest,
    Unimplemented,
    Unauthorized,
    AuthenticationHasBeenExpired,
    Forbidden,
    YouAreNotInInstance,
}
static SUTERA_STATUS_ERROR_MAP: Lazy<EnumMap<SuteraStatusError, [u8; 3]>> = Lazy::new(|| {
    enum_map! {
        SuteraStatusError::SchemaVersionNotSupported    => [0x10, 0x00, 0x00],
        SuteraStatusError::BadRequest                   => [0x10, 0x02, 0x00],
        SuteraStatusError::Unimplemented                => [0x10, 0x02, 0x01],
        SuteraStatusError::Unauthorized                 => [0x20, 0x00, 0x00],
        SuteraStatusError::AuthenticationHasBeenExpired => [0x20, 0x00, 0x01],
        SuteraStatusError::Forbidden                    => [0x20, 0x01, 0x00],
        SuteraStatusError::YouAreNotInInstance          => [0x20, 0x02, 0x00],
    }
});

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
pub enum SuteraStatusWarning {
    SchemaVersionNotExactlyMatched,
}
static SUTERA_STATUS_WARNING_MAP: Lazy<EnumMap<SuteraStatusWarning, [u8; 3]>> = Lazy::new(|| {
    enum_map! {
        SuteraStatusWarning::SchemaVersionNotExactlyMatched => [0x10, 0x00, 0x00],
    }
});

#[derive(Debug, PartialEq)]
pub enum SuteraStatus {
    Ok,
    Warning(SuteraStatusWarning),
    Error(SuteraStatusError),
}

impl ClockingFrame for SuteraStatus {
    type Context = ();

    const MIN_FRAME_SIZE: usize = 1;
    const MAX_FRAME_SIZE: usize = 4;

    async fn parse_frame_unchecked(
        cursor: &mut std::io::Cursor<&[u8]>,
        _ctx: &Self::Context,
    ) -> Option<Self> {
        match cursor.copy_to_bytes(1)[0] {
            0x00 => Some(Self::Ok),
            0x01 => {
                if cursor.remaining() < 3 {
                    return None;
                }
                let kind = [cursor.get_u8(), cursor.get_u8(), cursor.get_u8()];
                search_from_enum(*SUTERA_STATUS_WARNING_MAP, &kind).map(Self::Warning)
            }
            0x02 => {
                if cursor.remaining() < 3 {
                    return None;
                }
                let kind = [cursor.get_u8(), cursor.get_u8(), cursor.get_u8()];
                search_from_enum(*SUTERA_STATUS_ERROR_MAP, &kind).map(Self::Error)
            }
            _ => None,
        }
    }

    async fn write_frame<W: tokio::io::AsyncWriteExt + Unpin>(
        &self,
        stream: &mut W,
        _ctx: &Self::Context,
    ) -> std::io::Result<()> {
        match self {
            Self::Ok => stream.write_all(&[0x00]).await?,
            Self::Warning(w) => {
                stream.write_all(&[0x01]).await?;
                stream.write_all(&SUTERA_STATUS_WARNING_MAP[*w]).await?;
            }
            Self::Error(e) => {
                stream.write_all(&[0x02]).await?;
                stream.write_all(&SUTERA_STATUS_ERROR_MAP[*e]).await?;
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::clocking::traits::test_util::{
        test_clockingframe_mustfail, test_clockingframe_reflective,
    };

    use super::*;

    #[tokio::test]
    async fn clockingframe_sutera_status() {
        test_clockingframe_reflective(SuteraStatus::Ok, ()).await;
        test_clockingframe_reflective(
            SuteraStatus::Warning(SuteraStatusWarning::SchemaVersionNotExactlyMatched),
            (),
        )
        .await;
        test_clockingframe_reflective(SuteraStatus::Error(SuteraStatusError::BadRequest), ()).await;
        test_clockingframe_reflective(
            SuteraStatus::Error(SuteraStatusError::SchemaVersionNotSupported),
            (),
        )
        .await;
    }

    #[tokio::test]
    async fn clockingframe_sutera_status_prefixerr() {
        test_clockingframe_mustfail::<SuteraStatus>(&[0x02, 0x01, 0x00], &(), Some(1)).await;
        test_clockingframe_mustfail::<SuteraStatus>(&[0x03, 0x00, 0x00, 0x00], &(), Some(1)).await;
        test_clockingframe_mustfail::<SuteraStatus>(&[0x03, 0x01, 0x00, 0x00], &(), Some(1)).await;
    }

    #[tokio::test]
    async fn clockingframe_sutera_status_unexists() {
        test_clockingframe_mustfail::<SuteraStatus>(&[0x02, 0xFF, 0xFF, 0xFF], &(), Some(4)).await;
    }
}
