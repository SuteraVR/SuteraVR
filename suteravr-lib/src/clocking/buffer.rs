use crate::clocking::ClockingFrameUnit;
use crate::util::logger::Logger;
use crate::{debug, info, warn, error};

use super::oneshot_headers::OneshotHeader;
use super::sutera_header::SuteraHeader;
use super::sutera_status::SuteraStatus;
use super::traits::MessageAuthor;

pub struct FrameBuffer<T: Logger> {
    pub buffer: Vec<ClockingFrameUnit>,
    logger: T,
}

#[derive(Debug, Clone)]
pub enum ContentHeader {
    Oneshot(OneshotHeader)
}

#[derive(Debug, Clone)]
pub struct ReceivePayload {
    pub sutera_header: SuteraHeader,
    pub sutera_status: Option<SuteraStatus>,
    pub content_header: ContentHeader,
    pub payload: Vec<u8>,
}

impl<T: Logger> FrameBuffer<T> {
    #[inline]
    pub fn new(logger: T) -> Self {
        (&logger as &dyn Logger).write_debug("data".to_string());
        Self {
            buffer: Vec::with_capacity(4),
            logger,
        }
    }

    #[inline]
    fn push(&mut self, unit: ClockingFrameUnit) {
        debug!(self.logger, "Frame: {:?}", unit);
        self.buffer.push(unit);
    }

    #[inline]
    fn len(&self) -> usize {
        self.buffer.len()
    }

    #[inline]
    fn clear(&mut self) {
        self.buffer.clear()
    }

    #[inline]
    fn get_into(&self, index: usize) -> Option<ClockingFrameUnit> {
        self.buffer.get(index).cloned()
    }

    #[inline]
    pub fn append(
        &mut self,
        payload: ClockingFrameUnit,
        author: MessageAuthor,
    ) -> Option<ReceivePayload> {
        match payload {
            ClockingFrameUnit::SuteraStatus(_) => {
                error!(self.logger, "Unexpected SuteraStatus of ClockingConnection!");
                unreachable!();
            }
            ClockingFrameUnit::SuteraHeader(_) => {
                let len = self.len();
                if self.len() != 0 {
                    warn!(self.logger, "Skipped {} frame(s).", len);
                    self.clear();

                }
                self.push(payload);
            },
            ClockingFrameUnit::Unfragmented(c) => {
                warn!(self.logger, "Receive {} unfragmented byte(s)", c.len());
            },
            ClockingFrameUnit::Content(payload) => {
                let len = self.len();
                if len != match author {
                    MessageAuthor::Client => 2,
                    MessageAuthor::Server => 3,
                } {
                    warn!(self.logger, "Unexpected content, Skipped {} frame(s).", len);
                    self.clear();
                    return None;
                }
                let Some(ClockingFrameUnit::SuteraHeader(sutera_header)) = self.get_into(0) else {
                    return None;
                };
                let Ok(sutera_status) = (if author == MessageAuthor::Client { Ok(None) } else {
                    match self.get_into(1) {
                        Some(ClockingFrameUnit::SuteraStatus(sutera_status)) => Ok(Some(sutera_status)),
                        _ => {
                            Err(())
                        }
                    }
                }) else {
                    return None
                };
                match self.get_into(match author {
                    MessageAuthor::Client => 1,
                    MessageAuthor::Server => 2,
                }) {
                    Some(ClockingFrameUnit::OneshotHeaders(oneshot_header)) => {
                        let request = ReceivePayload {
                            sutera_header,
                            sutera_status,
                            content_header: ContentHeader::Oneshot(oneshot_header),
                            payload,
                        };
                        info!(self.logger, "Receive: {:?}", &request);
                        self.clear();
                        return Some(request);
                    },
                    Some(_) | None => {
                        return None;
                    },
                }
            }
            _ => {
                self.push(payload);
            }
        }
        None
    }
}
