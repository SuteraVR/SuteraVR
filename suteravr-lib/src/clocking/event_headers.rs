use bytes::Buf;
use derivative::Derivative;
use derive_new::new;
use enum_map::enum_map;
use enum_map::{Enum, EnumMap};
use once_cell::sync::Lazy;

use crate::util::search_from_enum;

use super::sutera_header::SuteraHeader;
use super::sutera_status::SuteraStatus;
use super::traits::{ClockingFrame, MessageAuthor};

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
#[allow(non_camel_case_types)]
pub enum EventTypes {
    Instance_PlayerJoined_Push,
    Instance_PlayerLeft_Push,
    Instance_PubPlayerMove_Pull,
    Instance_PushPlayerMove_Push,
    TextChat_ReceiveChatMessage_Push,
}

#[derive(Enum, PartialEq, Debug, Clone, Copy)]
pub enum EventDirection {
    Push,
    Pull,
}

static EVENT_TYPES_MAP: Lazy<EnumMap<EventTypes, [u8; EventHeader::MESSAGE_TYPE_DISTINCTOR_SIZE]>> =
    Lazy::new(|| {
        enum_map! {
            EventTypes::Instance_PlayerJoined_Push       => [0x00, 0x02, 0x00, 0x01],
            EventTypes::Instance_PlayerLeft_Push         => [0x00, 0x02, 0x00, 0x02],
            EventTypes::Instance_PubPlayerMove_Pull      => [0x00, 0x02, 0x01, 0x01],
            EventTypes::Instance_PushPlayerMove_Push     => [0x00, 0x02, 0x01, 0x02],
            EventTypes::TextChat_ReceiveChatMessage_Push => [0x00, 0x03, 0x00, 0x01],
        }
    });

pub static EVENT_TYPES_DIRECTION_MAP: Lazy<EnumMap<EventTypes, EventDirection>> = Lazy::new(|| {
    enum_map! {
        EventTypes::Instance_PlayerJoined_Push       => EventDirection::Push,
        EventTypes::Instance_PlayerLeft_Push         => EventDirection::Push,
        EventTypes::Instance_PubPlayerMove_Pull      => EventDirection::Pull,
        EventTypes::Instance_PushPlayerMove_Push     => EventDirection::Push,
        EventTypes::TextChat_ReceiveChatMessage_Push => EventDirection::Push,
    }
});

pub static EVENT_DIRECTION_MAP: Lazy<
    EnumMap<EventDirection, [u8; EventHeader::MESSAGE_DIRECTION_DISTINCTOR_SIZE]>,
> = Lazy::new(|| {
    enum_map! {
        EventDirection::Push => [0x02, 0x00],
        EventDirection::Pull => [0x03, 0x00],
    }
});

#[derive(Debug, PartialEq, Clone)]
pub struct EventHeader {
    pub direction: EventDirection,
    pub message_type: EventTypes,
}

impl EventHeader {
    pub const MESSAGE_DIRECTION_DISTINCTOR_SIZE: usize = 2;
    pub const MESSAGE_TYPE_DISTINCTOR_SIZE: usize = 4;
}

impl ClockingFrame for EventHeader {
    type Context = MessageAuthor;

    const MIN_FRAME_SIZE: usize =
        Self::MESSAGE_DIRECTION_DISTINCTOR_SIZE + Self::MESSAGE_TYPE_DISTINCTOR_SIZE;

    fn parse_frame_unchecked(
        cursor: &mut std::io::Cursor<&[u8]>,
        ctx: &Self::Context,
    ) -> Option<Self> {
        let dir = [cursor.get_u8(), cursor.get_u8()];
        match search_from_enum(*EVENT_DIRECTION_MAP, &dir) {
            Some(dir) => {
                let message_type = [
                    cursor.get_u8(),
                    cursor.get_u8(),
                    cursor.get_u8(),
                    cursor.get_u8(),
                ];
                let message_type = search_from_enum(*EVENT_TYPES_MAP, &message_type)?;
                if EVENT_TYPES_DIRECTION_MAP[message_type] != dir {
                    return None;
                }

                if dir
                    != match ctx {
                        MessageAuthor::Client => EventDirection::Pull,
                        MessageAuthor::Server => EventDirection::Push,
                    }
                {
                    return None;
                }
                Some(Self {
                    direction: dir,
                    message_type,
                })
            }
            None => None,
        }
    }

    async fn write_frame<W: tokio::io::AsyncWriteExt + Unpin>(
        &self,
        stream: &mut W,
        _ctx: &Self::Context,
    ) -> std::io::Result<()> {
        stream
            .write_all(&EVENT_DIRECTION_MAP[self.direction])
            .await?;
        stream
            .write_all(&EVENT_TYPES_MAP[self.message_type])
            .await?;
        Ok(())
    }
}

#[derive(Derivative, new)]
#[derivative(Debug)]
pub struct EventRequest {
    pub sutera_header: SuteraHeader,
    pub event_header: EventHeader,
    pub payload: Vec<u8>,
}
#[derive(Derivative, new)]
#[derivative(Debug)]
pub struct EventResponse {
    pub sutera_header: SuteraHeader,
    pub sutera_status: SuteraStatus,
    pub event_header: EventHeader,
    pub payload: Vec<u8>,
}
