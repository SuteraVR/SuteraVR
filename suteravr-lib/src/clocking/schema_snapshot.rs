#[cfg(test)]
mod test {
    use crate::clocking::{
        oneshot_headers::OneshotHeader, sutera_header::SuteraHeader, sutera_status::SuteraStatus,
        traits::ClockingFrame,
    };

    #[test]
    fn schema_snapshot() {
        insta::assert_debug_snapshot!(SuteraHeader::MIN_FRAME_SIZE);
        insta::assert_debug_snapshot!(SuteraHeader::MAX_FRAME_SIZE);

        insta::assert_debug_snapshot!(SuteraStatus::MIN_FRAME_SIZE);
        insta::assert_debug_snapshot!(SuteraStatus::MAX_FRAME_SIZE);

        insta::assert_debug_snapshot!(OneshotHeader::MIN_FRAME_SIZE);
        insta::assert_debug_snapshot!(OneshotHeader::MAX_FRAME_SIZE);
    }
}
