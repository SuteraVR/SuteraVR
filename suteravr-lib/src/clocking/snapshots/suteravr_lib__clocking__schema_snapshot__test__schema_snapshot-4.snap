---
source: src/clocking/schema_snapshot.rs
expression: "SuteraStatus::MAX_FRAME_SIZE"
---
4
